core/runtime/op_driver/erased_future.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+use std::future::Future;
+use std::marker::PhantomData;
+use std::marker::PhantomPinned;
+use std::mem::MaybeUninit;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+#[repr(C, align(16))]
+pub struct TypeErased<const MAX_SIZE: usize> {
+  memory: [MaybeUninit<u8>; MAX_SIZE],
+  drop: fn(this: *mut ()),
+  _unpin: PhantomPinned,
+}
+
+impl<const MAX_SIZE: usize> TypeErased<MAX_SIZE> {
+  pub unsafe fn take<R: 'static>(self) -> R {
+    assert!(
+      std::mem::size_of::<R>() <= std::mem::size_of_val(&self.memory),
+      "Invalid size for {}: {} (max {})",
+      std::any::type_name::<R>(),
+      std::mem::size_of::<R>(),
+      std::mem::size_of_val(&self.memory)
+    );
+    assert!(
+      std::mem::align_of::<R>() <= std::mem::align_of_val(&self),
+      "Invalid alignment for {}: {} (max {})",
+      std::any::type_name::<R>(),
+      std::mem::align_of::<R>(),
+      std::mem::align_of_val(&self)
+    );
+    let r = std::ptr::read(self.memory.as_ptr() as _);
+    std::mem::forget(self);
+    r
+  }
+
+  pub fn raw_ptr(&mut self) -> *mut () {
+    self.memory.as_mut_ptr() as _
+  }
+
+  pub fn new<R>(r: R) -> Self {
+    let mut new = Self {
+      memory: [MaybeUninit::uninit(); MAX_SIZE],
+      drop: |this| unsafe { std::ptr::drop_in_place::<R>(this as _) },
+      _unpin: PhantomPinned,
+    };
+    assert!(
+      std::mem::size_of::<R>() <= std::mem::size_of_val(&new.memory),
+      "Invalid size for {}: {} (max {})",
+      std::any::type_name::<R>(),
+      std::mem::size_of::<R>(),
+      std::mem::size_of_val(&new.memory)
+    );
+    assert!(
+      std::mem::align_of::<R>() <= std::mem::align_of_val(&new),
+      "Invalid alignment for {}: {} (max {})",
+      std::any::type_name::<R>(),
+      std::mem::align_of::<R>(),
+      std::mem::align_of_val(&new),
+    );
+    unsafe { std::ptr::write(new.memory.as_mut_ptr() as *mut R, r) };
+    new
+  }
+}
+
+impl<const MAX_SIZE: usize> Drop for TypeErased<MAX_SIZE> {
+  fn drop(&mut self) {
+    (self.drop)(self.raw_ptr())
+  }
+}
+
+pub struct ErasedFuture<const MAX_SIZE: usize, Output> {
+  erased: TypeErased<MAX_SIZE>,
+  poll: unsafe fn(
+    this: Pin<&mut TypeErased<MAX_SIZE>>,
+    cx: &mut Context,
+  ) -> Poll<Output>,
+  _output: PhantomData<Output>,
+}
```

Comment:
A better name would be `StackFuture`.

But I think we should use the `stackfuture` crate from Microsoft https://github.com/microsoft/stackfuture - the implementation looks similar. `stackfuture` also has compile-time assertions when future does not fit.

Reply:
Thanks for the pointer to the `stackfuture` crate, I haven't seen that but it could be useful in other places.

In our case we cannot use the crate because we also need the `TypeErased` struct to erase the value types as well.

We don't need compile-time size assertions because we do a compile-time check and allocate on the heap if the future or return value doesn't fit (see the ValueLarge enum member, for instance).

`stackfuture` isn't really a great name for this as these don't live in the stack -- they live in the Arena. They are copied around like value types only at the very beginning, after which they are pinned in place inside the Arena.

Reply:
I see, sounds good.